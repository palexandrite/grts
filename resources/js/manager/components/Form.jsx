import React from "react";
import { Link, withRouter } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import PropTypes from "prop-types";

import Spinner from "./Spinner";

// const fetchBefore = async (url, params) => {
//     try {
//         const response = await fetch(url, params);
//         const responseJson = await response.json();

//         return responseJson;
//     } catch ( error ) {
//         console.error("Pipe all hands on deck! We've got an error with the response", error)
//     }
// };

class Form extends React.Component
{
    static propTypes = {
        match: PropTypes.object.isRequired,
        location: PropTypes.object.isRequired,
        history: PropTypes.object.isRequired
    };

    constructor(props)
    {
        super(props);
        this.state = {
            errors: null,
            item: {},
            isCreatePage: false,
            isRendered: false,
            successSubmitMessage: null,
            serverErrorMessage: null,
        };

        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleChange = this.handleChange.bind(this);
        this.prepareItem = this.prepareItem.bind(this);
        this.invalidInputCssClass = "is-invalid";
        this.submitButtonRef = React.createRef();
        this.submitTextRef = React.createRef();
        this.submitSpinnerRef = React.createRef();
    }

    componentWillUnmount()
    {
        this.setState({
            item: {},
        });
    }

    getCookie( name )
    {
        let matches = document.cookie.match(new RegExp(
            "(?:^|; )" + name.replace(/([\.$?*|{}\(\)\[\]\\\/\+^])/g, '\\$1') + "=([^;]*)"
        ));
        return matches ? decodeURIComponent(matches[1]) : undefined;
    }

    componentDidMount()
    {
        if ( this.props.match.params.id ) {

            const token = this.getCookie("atoken");
            let url = "/api/manager/" + this.props.model + "/show",
                params = {
                    method: "POST",
                    headers: {
                        "Content-Type": "application/json",
                        "Authorization": "Bearer " + token,
                        "X-Requested-With": "XMLHttpRequest",
                        // "X-CSRF-Token": document.querySelector("meta[name=csrf-token]").content
                    },
                    body: JSON.stringify({
                        item: this.props.match.params.id
                    }),
                };
    
            fetch( url, params )
                .then(response => response.json())
                .then(data => {
                    this.setState({
                        item: this.prepareItem( data ),
                        isRendered: true
                    });
                })
                .catch(error => console.error("Pipe all hands on deck! We've got an error with the response", error));

        } else if ( this.props.match.url ) {

            let match = this.props.match.url.match( new RegExp("\/create$", "i") ) || [];
            if ( match.length > 0 ) {
                this.setState({
                    item: this.prepareItem( null ),
                    isCreatePage: true,
                    isRendered: true,
                });
            }

        } else {
            throw new Error("Something went wrong related to the URL!");
        }
    }

    prepareItem( item )
    {
        if ( item !== null ) {
            item.password = "";
            return item;
        } else {
            let $item = {};

            this.props.fields.map((value, index) => {
                if ( index === 0 ) return;
                $item[ value.inputElement.name ] = value.value;
            });

            return $item;
        }
    }

    handleSubmit( e )
    {
        e.preventDefault();

        let submitButton = this.submitButtonRef.current;
        let submitText = this.submitTextRef.current;
        let submitSpinner = this.submitSpinnerRef.current;

        submitButton.disabled = true;
        submitText.classList.toggle("d-none");
        submitSpinner.classList.toggle("d-none");

        const token = this.getCookie("atoken");
        let url = "/api/manager/" + this.props.model + "/" + this.props.url,
            params = {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                    "Authorization": "Bearer " + token,
                    "X-Requested-With": "XMLHttpRequest",
                    // "X-CSRF-Token": document.querySelector("meta[name=csrf-token]").content
                },
                body: JSON.stringify(this.state.item),
            };

        fetch( url, params )
            .then(response => response.json() )
            .then(data => {

                console.log('this is a response');
                console.dir(data);

                submitButton.disabled = false;
                submitText.classList.toggle("d-none");
                submitSpinner.classList.toggle("d-none");

                if (data.exception) {
                    this.setState({
                        serverErrorMessage: "Suddenly, some server error was occurred"
                    });
                    throw new Error("Oh, holy cow! We've caught some server error on the board!");
                }

                if ( data.errors ) {
                    this.setState({errors: data.errors});
                } else if ( data.success ) {
                    this.setState({successSubmitMessage: data.success});
                }
            })
            .catch(error => console.error("Pipe all hands on the deck! We've got an error with the submit.", error));
    }

    handleChange( e )
    {
        const target = e.target;
        const value = target.type === "checkbox" ? target.checked : target.value;
        const name = target.name;

        this.setState((prevState) => {
            let prevItem = prevState.item;
            prevItem[ name ] = value;
            return {
                item: prevItem,
                errors: null,
                serverErrorMessage: null,
                successSubmitMessage: null,
            };
        });
    }

    getLabelElement( object )
    {
        let $label = false;

        if ( object.labelElement ) {

            const { text, ...attributes } = object.labelElement;

            $label = (
                <label { ...attributes }>{ text }</label>
            );

        } else if ( object.spanElement ) {

            const { ...attributes } = object.spanElement;

            $label = (
                <span { ...attributes }>{ "ID: " + object.value }</span>
            );

        }

        return $label;
    }

    getAutocompleteForInput( inputType )
    {
        let autocomplete;

        if ( inputType === "password" ) {
            autocomplete = this.state.isCreatePage ?  "new-password" : "current-password";
        } else if ( inputType === "email" ) {
            autocomplete = "email";
        }

        return autocomplete !== null ? autocomplete : undefined;
    }

    getCurrentValidation( inputName )
    {
        let validationErrors = this.state.errors,
            element = {};

        if ( validationErrors !== null && validationErrors[ inputName ] ) {
            element.valElement = (
                <div className="invalid-feedback">
                    { validationErrors[ inputName ].join("; ") }
                </div>
            );
            element.valCssClass = this.invalidInputCssClass;
        } else {
            element = {
                valElement: null,
                valCssClass: ""
            };
        }

        return element;
    }

    getCurrentClassList( classList, validationClass )
    {
        let current = classList.search( this.invalidInputCssClass );

        if ( validationClass !== "" ) {
            if ( current == -1 ) {
                classList += " " + validationClass;
            } 
        } else {
            if ( current > -1 ) {
                classList = classList.replace( " " + this.invalidInputCssClass, "" );
            }
        }

        return classList;
    }

    getInputElement( object )
    {
        let $inputElement = false,
            $checkboxValElement = null;

        if ( object.inputElement ) {

            let { name } = object.inputElement;

            let { valElement, valCssClass } = this.getCurrentValidation( name );

            if ( object.inputElement.className ) {
                let classList = object.inputElement.className;
                object.inputElement.className = this.getCurrentClassList( classList, valCssClass );
            } 

            let autocomplete = this.getAutocompleteForInput( object.inputElement.type );

            if ( object.inputElement.type === "hidden" ) {
                $inputElement = (
                    <input { ...object.inputElement } value={ object.value } />
                );
            } else if (
                object.inputElement.type === "checkbox" ||
                object.inputElement.type === "radio"
            ) {
                $inputElement = (
                    <input
                        { ...object.inputElement } 
                        defaultChecked={ object.value === "Active" }
                        onChange={ this.handleChange } />
                );
                $checkboxValElement = valElement;
            } else {
                $inputElement = (
                    <>
                    <input
                        { ...object.inputElement }
                        autoComplete={ autocomplete }
                        defaultValue={ object.value }
                        onChange={ this.handleChange } />
                    { valElement }
                    </>
                );
            }
        }

        return [ $inputElement, $checkboxValElement ];
    }

    getLayoutOfElement( object, label, input, valElementOfCheckbox )
    {
        let { beginLayoutWith } = object;

        if ( beginLayoutWith !== undefined && beginLayoutWith === "input" ) {
            return (
                <>{ input }{ label }{ valElementOfCheckbox }</>
            );
        }

        return (
            <>{ label }{ input }</>
        );
    }

    getFormElements()
    {
        let $output = [];

        if ( Object.keys( this.props.fields ).length > 0 ) {

            this.props.fields.map((value, index) => {

                if ( this.state.isCreatePage && index === 0 ) {
                    return;
                } else if (! this.state.isCreatePage ) {
                    for ( let prop in this.state.item )
                    {
                        if ( this.state.item.hasOwnProperty( prop ) ) {
                            if ( value.inputElement.name === prop ) {
                                value.value = this.state.item[ prop ];
                                break;
                            }
                        }
                    }
                }

                let $label = this.getLabelElement( value );

                let [ $inputElement, $valCheckboxElement ] = this.getInputElement( value );

                let $elementLayout = this.getLayoutOfElement( value, $label, $inputElement, $valCheckboxElement );

                if ( value.wrapperElement ) {
                    $output.push(
                        <div key={ index } { ...value.wrapperElement }>
                            { $elementLayout }
                        </div>
                    );
                }
            });
        }

        return $output;
    }

    renderForm()
    {
        let { successSubmitMessage, serverErrorMessage } = this.state;

        return (
            <form onSubmit={ this.handleSubmit }>

                { this.getFormElements() }

                <button 
                    ref={ this.submitButtonRef }
                    type="submit"
                    className="btn btn-primary mr-5"
                >
                    <span ref={ this.submitTextRef }>Submit</span>
                    <span className="d-none" ref={ this.submitSpinnerRef }>
                        <span className="spinner-border spinner-border-sm" role="status" aria-hidden="true"></span>
                        Loading...
                    </span>
                </button>

                <span className="d-block d-sm-inline pt-1 ms-sm-3 text-success">
                    { successSubmitMessage }
                </span>
                <span className="d-block d-sm-inline pt-1 ms-sm-3 text-danger">
                    { serverErrorMessage }
                </span>

            </form>
        );
    }

    /*
     * The main method of the object
    */
    render()
    {
        const { isRendered } = this.state;
        const linkTo = "/manager/" + this.props.model;

        return (
            <div className="card">
                <div className="card-header">
                    <Link 
                        to={ linkTo }
                        className="btn btn-link"
                        data-object="expandable"
                    >
                        <FontAwesomeIcon 
                            icon={ ["fas", "chevron-left"] } 
                            size="1x" />
                        <span>
                            Back to all
                        </span>
                    </Link>
                    <span className="border-start ps-sm-2">{ this.props.currentText }</span>
                </div>
                <div className="card-body">

                    { isRendered ? this.renderForm() : <Spinner /> }
                    
                </div>
            </div>
        );
    }
}

export default withRouter( Form );