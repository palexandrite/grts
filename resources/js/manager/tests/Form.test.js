import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import { MemoryRouter, Route } from "react-router-dom";
import renderer from 'react-test-renderer'; 
import pretty from "pretty";

import Form from "../components/Form";
import UserFormFields from "../pages/users/FormFields";
import FontAwesomeRegister from "../fontawesome";

FontAwesomeRegister();

let container = null;

beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement("div");
    document.body.appendChild(container);
});

afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
});

it("The user form is loading", () => {
    // The test without componentDidMount
    const component = renderer.create(
        <MemoryRouter>
            <Form
                model="users" 
                url="create"
                fields={ UserFormFields() }
                currentText="Create an user" />
        </MemoryRouter>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it("The user creation form was rendered", () => {
    act(() => {
        render(
            <MemoryRouter initialEntries={['/manager/users/create']} initialIndex={ 1 }>
                <Route path="/manager/users/create">
                    <Form
                        model="users" 
                        url="create"
                        fields={ UserFormFields() }
                        currentText="Create an user" />
                </Route>
            </MemoryRouter>, 
            container
        );
    });
    expect(pretty(container.innerHTML)).toMatchSnapshot();
});