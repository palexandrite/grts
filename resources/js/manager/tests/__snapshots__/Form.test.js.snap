// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The user creation form was rendered 1`] = `
"<div class=\\"card\\">
  <div class=\\"card-header\\"><a class=\\"btn btn-link\\" data-object=\\"expandable\\" href=\\"/manager/users\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"chevron-left\\" class=\\"svg-inline--fa fa-chevron-left fa-w-10 fa-1x \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 320 512\\">
        <path fill=\\"currentColor\\" d=\\"M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z\\"></path>
      </svg><span>Back to all</span></a><span class=\\"border-start ps-sm-2\\">Create an user</span></div>
  <div class=\\"card-body\\">
    <form>
      <div class=\\"mb-3\\"><label for=\\"first-name\\" class=\\"form-label\\">First Name</label><input id=\\"first-name\\" name=\\"first_name\\" type=\\"text\\" class=\\"form-control\\" placeholder=\\"Type a first name...\\" value=\\"\\"></div>
      <div class=\\"mb-3\\"><label for=\\"last-name\\" class=\\"form-label\\">Last Name</label><input id=\\"last-name\\" name=\\"last_name\\" type=\\"text\\" class=\\"form-control\\" placeholder=\\"Type a last name...\\" value=\\"\\"></div>
      <div class=\\"mb-3\\"><label for=\\"email\\" class=\\"form-label\\">Email</label><input id=\\"email\\" name=\\"email\\" type=\\"email\\" class=\\"form-control\\" placeholder=\\"Type an email...\\" autocomplete=\\"email\\" value=\\"\\"></div>
      <div class=\\"mb-3\\"><label for=\\"password\\" class=\\"form-label\\">Password</label><input id=\\"password\\" name=\\"password\\" type=\\"password\\" class=\\"form-control\\" placeholder=\\"Set a password...\\" autocomplete=\\"new-password\\" value=\\"\\"></div>
      <div class=\\"form-check form-switch my-4\\"><input id=\\"status\\" name=\\"status\\" type=\\"checkbox\\" class=\\"form-check-input\\"><label for=\\"status\\" class=\\"form-check-label\\">Does the user has the approved email?</label></div><button type=\\"submit\\" class=\\"btn btn-primary mr-5\\"><span>Submit</span><span class=\\"d-none\\"><span class=\\"spinner-border spinner-border-sm\\" role=\\"status\\" aria-hidden=\\"true\\"></span>Loading...</span></button><span class=\\"d-block d-sm-inline pt-1 ms-sm-3 text-success\\"></span><span class=\\"d-block d-sm-inline pt-1 ms-sm-3 text-danger\\"></span>
    </form>
  </div>
</div>"
`;

exports[`The user form is loading 1`] = `
<div
  className="card"
>
  <div
    className="card-header"
  >
    <a
      className="btn btn-link"
      data-object="expandable"
      href="/manager/users"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-chevron-left fa-w-10 fa-1x "
        data-icon="chevron-left"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      <span>
        Back to all
      </span>
    </a>
    <span
      className="border-start ps-sm-2"
    >
      Create an user
    </span>
  </div>
  <div
    className="card-body"
  >
    <div
      className="d-flex align-items-center text-secondary"
    >
      <strong>
        Loading...
      </strong>
      <div
        aria-hidden="true"
        className="spinner-border ms-auto"
        role="status"
      />
    </div>
  </div>
</div>
`;
